File: Python/099-recover-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def recoverTree(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        self.first, self.second, self.pre = None, None, None
        #中序遍历二叉树，找到递减的节点，如果有一个交换他们相邻的，如果有两个，交换这两个
        def inorder(root: TreeNode) -> None:
            if root == None:
                return None
            if root.left != None:
                inorder(root.left)
            if self.pre != None and self.pre.val > root.val:
                if self.first == None:
                    self.first = self.pre
                self.second = root
            self.pre = root
            if root.right != None:
                inorder(root.right)
        
        if root == None:
            return None
        inorder(root)
        self.first.val, self.second.val = self.second.val, self.first.val

File: Go/099-recover-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

 var (
    first *TreeNode
    second *TreeNode
    pre *TreeNode
)

//中序遍历二叉树，找到递减的节点，如果有一个交换他们相邻的，如果有两个，交换这两个
func inorder(root *TreeNode) {
    if root == nil {
        return
    }
    if root.Left != nil {
        inorder(root.Left)
    }
    if pre != nil && pre.Val > root.Val {
        if first == nil {
            first = pre
        }
        second = root
    }
    pre = root
    if root.Right != nil {
        inorder(root.Right)
    }
}

func recoverTree(root *TreeNode)  {
    first, second, pre = nil, nil, nil
    if root == nil {
        return
    }
    inorder(root)
    first.Val, second.Val = second.Val, first.Val
}